Bring Energenie repo up to date with new device additions

whaleygeek opened the pull request:
This PR rolls together a large batch of changes that were originally done on the device_classes branch, including:
- Device classes work integrated and working fine.
- All demo programs rewritten to use the new simpler device class abstractions.
- Auto device discovery feature added.
- setup_tool.py written to ease the setup process.
- A whole load more new device classes were added and tested against sample product loaned to me by Energenie.
- Much better getting started instructions in the readme file.

whaleygeek committed:
Started a doc to spec and design the device classes

whaleygeek committed:
Done a requirements and design pass of the device classes idea

whaleygeek committed:
Started to flesh out design thinking of device classes

whaleygeek committed:
Started to flesh out device classes proposed interfaces

whaleygeek committed:
Method signatures for device classes added

whaleygeek committed:
Some ideas about device registration added

whaleygeek committed:
Initial design for enhanced persistent registry done

whaleygeek committed:
Method signatures for Registry sketched out

whaleygeek committed:
Started to implement device classes code.
Simple mocking interface now working.

whaleygeek committed:
Documented requirement for a simple persistent data store for the Registry

whaleygeek committed:
Mock registry store created, auto_create is working

whaleygeek committed:
Capabilities of devices turned into flags, with query method in base class

whaleygeek committed:
Initial design of a message Router

whaleygeek committed:
Started to plan out the tx/rx pipeline for payloads and message buffers

whaleygeek committed:
Renamed encoder as TwoBit, as it encodes two bits per byte.
This is the encoding protocol used with legacy OOK devices

whaleygeek committed:
More design on the tx/rx and air_interface pipeline

whaleygeek committed:
Placeholder added to architecture for radio interface adaptors

whaleygeek committed:
Spec'ed out responsibilities of OnAir interfaces

whaleygeek committed:
Method signatures defined for OnAir adaptors

whaleygeek committed:
Implementation checklist added for OnAir

whaleygeek committed:
Most of the OnAir interface coded or spec'ed now

whaleygeek committed:
LegacyDevice tx path commissioned end to end

whaleygeek committed:
Commissioned a tx path for MiHome adaptor Plus.
This is a temporary path, but it generates the correct message

whaleygeek committed:
Deprecated Messages due to import race with Devices.
Template messages (which are cloned with create_message) now in Devices.

whaleygeek committed:
Spec'ed interface for OOK decoder

whaleygeek committed:
Started to flesh out the MiHome receive pipeline

whaleygeek committed:
Poked a synthetic switch on message down the MIHO005 receive pipeline

whaleygeek committed:
More representative synthetic data added to FSK receive test

whaleygeek committed:
Moved test harnesses into separate _test.py modules

whaleygeek committed:
Major tidy up of test harness and lifecycle methods

whaleygeek committed:
Improved separation of test harness messages

whaleygeek committed:
Started to think through a better alterMessage
characters that can be used in an identifier need checking

whaleygeek committed:
Started to think through a better message abstraction

whaleygeek committed:
Message.get() implemented and tested with paramid keying

whaleygeek committed:
Started writing unittest harness for new OpenThings.Message abstraction

whaleygeek committed:
Clarified test cases for OpenThings.Message()

whaleygeek committed:
Refinements to different parameter formats

whaleygeek committed:
Characterised usage modes of OpenThings.Message better

whaleygeek committed:
OpenThings.Message test harness keyed paths tested and working

whaleygeek committed:
Tested all attribute read services in OpenThings.Message

whaleygeek committed:
setattr tests done on OpenThings.Message

whaleygeek committed:
OpenThings.Message now tests attribute set and get.
Added decorators for @test_0 and @test_1 to turn tests on and off selectively

whaleygeek committed:
PARAM_NAME key syntax for OpenThings.Message [set] tested ok

whaleygeek committed:
OpenThings.Message tests now completed

whaleygeek committed:
Marked with ##HERE markers the refactoring required for new Message()
Disabled alterMessage showMessage and createMessage to force this issue

whaleygeek committed:
Refactored usages of raw message to use Message()
Removed all usages of showMessage, alterMessage

whaleygeek committed:
Code now ready for regression testing with new Message() wrappers added

whaleygeek committed:
Ran all test cases again (no hardware), fixed a few issues.
Need to still run this up on the hardware to test properly

whaleygeek committed:
Improved debug output for testing

whaleygeek committed:
Updated TODO list for the next phase of work on device_classes branch
Specifically re-testing samples, then pushing OO classes up into app.

Commit:
Tested on Raspberry Pi, all 4 testers still work with new Message()

whaleygeek committed:
legacy.py rewritten to use new device class interface

whaleygeek committed:
switch.py renamed as control_mihome.py, rewritten for new OO interface

whaleygeek committed:
control_both.py working in simulation

whaleygeek committed:
Router ready to be tied into a loop()

whaleygeek committed:
energenie.loop() written with FSK receiver connected to fsk_router

whaleygeek committed:
Updated TODO list with next steps for receive pipeline

whaleygeek committed:
Synthetic receive working in monitor_mihome.py

whaleygeek committed:
Fixed bug where sensorid was always zero in csv log

whaleygeek committed:
Neater callback mechanism for device class instances to use.
This new architecture ensures default message processing always happens

whaleygeek committed:
Router unknown message handling callback registration is cleaner

whaleygeek committed:
Default incoming message callback added to router, for logging/debug

whaleygeek committed:
Can now construct a device class instance from its product_id

whaleygeek committed:
Updated plan for next steps

whaleygeek committed:
Design for discovery agents done

whaleygeek committed:
Reviewed discovery user requirements, updated todo list

whaleygeek committed:
Framework of discovery methods in place, ready to TDD

whaleygeek committed:
Initial outer testing of 4 discovery modes done ok

whaleygeek committed:
Auto discovery tested and working with accept

whaleygeek committed:
Most test cases for discovery modes implemented and passing

whaleygeek committed:
Join_ack working on discovery mode autojoin

whaleygeek committed:
Set up test case for the final discovery test.
Test not run yet, and code is missing, marked with ####HERE####

whaleygeek committed:
Registry_test all test cases now pass for discovery

whaleygeek committed:
Factory method added to MiHomeDevice to generate a join_req for testing

whaleygeek committed:
monitor_mihome now works with auto join and switch toggling

whaleygeek committed:
Thought through next set of demos

whaleygeek committed:
Design for Registry data store (KVS) done

whaleygeek committed:
Pseudo code for KVS written, ready to be coded up

whaleygeek committed:
Design of persistent registry store improved after a re-think

whaleygeek committed:
Test harness for new Registry and KVS in place

whaleygeek committed:
Beginnings of a persistent version of a class is appearing

whaleygeek committed:
Managed to persist device configuration to registry.kvs

whaleygeek committed:
Moved KVS to separate module with separate test harness.
Will test this in isolation first before integrating with Registry

whaleygeek committed:
Test harness and functionality for KVS defined

whaleygeek committed:
KVS tests for non persistent KVS all pass

whaleygeek committed:
KVS tested in non persisted mode and works.
KVS mostly tested in persisted mode, but more code to write and test

whaleygeek committed:
KVS.remove() test now passes

whaleygeek committed:
kvs remove functionality working

whaleygeek committed:
Factory creation in ADD is now working

whaleygeek committed:
KVS.DEL() passing tests

whaleygeek committed:
KVS all test cases pass

whaleygeek committed:
Most of Registry_test completed.
Just need to write code that passes the 'rx routes created' test

whaleygeek committed:
Just the route creator to test now

whaleygeek committed:
All registry tests now pass.
Requirements list reviewed.
TODO list updated for next steps

whaleygeek committed:
Devices_test is now a unittest

whaleygeek committed:
Receive counter implemented and tested

whaleygeek committed:
Specified the demos that we want to write as part of the device classes testing

whaleygeek committed:
Defined requirements for demo apps including setup tool menu

whaleygeek committed:
setup_tool main menu written and working

whaleygeek committed:
Design for all setup_tool actions done

whaleygeek committed:
About half of the setup tool is written and roughly working

whaleygeek committed:
setup_tool now written, ready to start testing

whaleygeek committed:
setup_tool finished, ready for testing

whaleygeek committed:
setup_tool tested in simulation and works OK

whaleygeek committed:
control_legacy.py simplified, assuming learning done by setup_tool.py

whaleygeek committed:
control_mihome.py recoded to use remembered registry

whaleygeek committed:
Removed control_both.py, as control_any.py will do any sockets now

whaleygeek committed:
Renamed a lot of the control_ demos, due to device agnostic behaviour

whaleygeek committed:
Clarified control_any_noreg.py by adding a MIHO005 example

whaleygeek committed:
Iterator added to registry

whaleygeek committed:
Improved registry listing by using iterator

whaleygeek committed:
Standalone discovery example written.
Not able to test until real devices added

whaleygeek committed:
Energy monitor example written.
Not tested yet, as need to test with real devices

whaleygeek committed:
auto variable create example added.
Tested in simulation, seems to work, but need to re-test on hardware

whaleygeek committed:
Moved setup code from global scope of Registry to energenie.init()
This makes it possible to use the registry in test cases

whaleygeek committed:
Rationalised all comment styles.
The double hash (##) is reserved for commented out lines of code.
The single hash (#) is reserved for normal comments.

whaleygeek committed:
All testers except Registry_test now running all tests

whaleygeek committed:
All test cases now run fine.
There is no capture/compare at the moment, so verification is manual

Commit:
Tested setup_tool.py on real hardware - works OK
Some tweaks required but now works fine

Commit:
Tested demos on real hardware. Works fine.
Observation: handling of incoming messages does not check registry,
it only checks the router, so if you don't have a device object
bound to it, it will be seen as unknown. Needs fixing, but not a high
priority.

Commit:
Tested and tweaked for Python 3 support

whaleygeek committed:
New tests documented.
All tests done and pass

whaleygeek committed:
Merge pull request #61 from whaleygeek/device_classes

Device classes

whaleygeek committed:
Updated README.md to bring in line with latest release

whaleygeek committed:
Removed sample programs and moved to issues in the issue log
These are not ready to be released yet and are at early ideas stage.

whaleygeek committed:
Reviewed TODO list from the device_classes branch
The remaining items will be moved to an issue for later mop up

whaleygeek committed:
working notes file device_classes_branch.txt moved over to an issue

whaleygeek committed:
Now runs on Raspberry Pi 3 also

whaleygeek committed:
Spec'ed out work required to add the new devices

whaleygeek committed:
Documented messages from 3 new devices

whaleygeek committed:
Added MIHO033 OpenSensor to Devices.py

Commit:
Support for MIHO003 OpenSensor tested and working

whaleygeek committed:
Added support for MIHO032 MotionSensor

Commit:
Tested MIHO032 PIR sensor, works fine

whaleygeek committed:
Reviewed TODO items for new device support

whaleygeek committed:
Merge pull request #73 from whaleygeek/new_devices_june16

New devices june16

whaleygeek committed:
Documented work required to get the MiHome Light integrated

whaleygeek committed:
Clearly documented plan of steps to implement the change

whaleygeek committed:
Device.config is now Device.radio_config
This is to prevent confusion with get_config and get_radio_config

whaleygeek committed:
radio_config now making it through to OnAir.TwoBitInterface.send()

whaleygeek committed:
Now passes and uses inner_times override to radio driver
MIHO008 white_light working in simulation
and the ENER002 still uses the older inner_times=8

whaleygeek committed:
inner_times now correctly set on both interfaces.
Tested in simulation

whaleygeek committed:
Added support for MIHO014 in line relay.
Tested in simulation

whaleygeek committed:
Added in support for other devices like control and monitor only

Commit:
Tested with new devices on Raspberry Pi. All work OK

whaleygeek committed:
Branch ready for merging back to master

whaleygeek committed:
Merge pull request #74 from whaleygeek/ook_radio_config

Ook radio config

whaleygeek committed:
Updated README with better links to places to get the code from

Commit:
Tested MIHO014 relay on Raspberry pi - works fine

whaleygeek commented:
@LibanO this pull request rolls together a whole load of work done in May, that makes the code much easier to engage with for users. There are many out of the box demo programs included, as well as a nice setup_tool.py. Can you work through it and test it, and when you are ready, can you 'merge pull request' to make this set of files current in your fork of the repo, thanks!

ajtucker committed:
Small fixes for Python 3

whaleygeek committed:
Merge pull request #79 from ajtucker/master

Small fixes for Python 3

whaleygeek committed:
Update README.md

whaleygeek committed:
Update README.md

whaleygeek committed:
Update README.md

whaleygeek committed:
Update README.md

genestealer committed:
Updated MIHome-Light-User-Guide pdf

whaleygeek committed:
Update Devices.py to list new MIHO069 Wall Thermostat

TWilkin committed:
Added setup.py build for pyenergenie.

webash commented:
Hey there - I notice that this was submitted in Sep 2016 (2 years ago!) and the repo's code hasn't been touched in 3 years... do you think this project (and possibly also your contributions?) are dead?

whaleygeek commented:
Hi @webash Energenie may have a view on this - I think there is a favour over stability here rather than change, but I'll let them answer that question.

If you ask the same question at the root repo (github.com/whaleygeek/pyenergenie) I can give you my personal perspective of the progress and future of work that is going on over there.

Achronite committed:
Fixes to pyenergenie C library

3 Fixes merged from my fork.
1) Allow to run as standard user (not root)
2) Flash LEDs during init
3) Place radio in standby on finish

Achronite committed:
Create .gitignore

Added .pyc files to .gitignore

whaleygeek committed:
Merge pull request #101 from Achronite/master

Run without root, radio standby on finish, flash LEDs during init

whaleygeek committed:
Merge pull request #89 from genestealer/master

Updated MIHome-Light-User-Guide pdf

jmigreen committed:
Run from anywhere

Makes the registry path dynamic so scripts can reference it from any working directory. Starts line 48.

whaleygeek committed:
Merge pull request #104 from jmigreen/patch-1

Always uses the registry.kvs stored in the pyenergenie project (rather than using registry.kvs in the cwd).

byxor committed:
Add missing import for 'sys'

whaleygeek committed:
Merge pull request #105 from byxor/master

Add missing import for 'sys'
@jonathangroves confirmed that this has been seen working, so I think that's good enough info to merge to master.

DrOwl commented:
Come on Energenie, lets have some progress.

TWilkin committed:
Update setup.py

Added handling for newer pip versions.

TWilkin committed:
Merging changes in upstream to fork

TWilkin committed:
Update setup.py

Fixed package path.

Commit:
Enable Raspberry Pi build on platforms without bash

TWilkin committed:
Fixing requirements import

TWilkin committed:
Fixing imports when used from wheel

whaleygeek committed:
Wrapped setup_tool with a crash check

whaleygeek committed:
Added message relating to sudo

whaleygeek committed:
Clarified setup_tool error handling workflow

whaleygeek committed:
Wrote a simple 2-gang light switch test harness

whaleygeek committed:
Parameterised 2g tester

whaleygeek committed:
Hardened GPIO cleanup

whaleygeek committed:
Best paramaters for 2gand switch, burst size about 1.1s, holdoff 3s

Commit:
Added support for MIHO010, MIHO075, MIHO076, MIHO077 & MIHO87 Single Gang Dimmer Switches.

whaleygeek committed:
Merge pull request #127 from Shinbar/SingleGangDimmerSupport

Added support for Single Gang Dimmer Switches

whaleygeek committed:
Removed future import (does not work on some python versions)
Created a setup_tool bash script that runs as root with correct python.

whaleygeek committed:
Updated README with a link to the external EV1527 support.
Really did remove the 'future' for True, as it doesn't work in my build.

whaleygeek committed:
Replaced reference to setup_tool.py with ./setup_tool

whaleygeek committed:
Rationalised cleanup in setup_tool and test_2g so GPIO always cleaned up.

whaleygeek committed:
Merge pull request #122 from BobSammers/master

Enable Raspberry Pi build on platforms without bash

whaleygeek committed:
Merge branch 'master' into master

whaleygeek committed:
Merge pull request #124 from TWilkin/master

Python 3 and wheel building setup

whaleygeek committed:
Updated table of product names

whaleygeek committed:
Added slightly more standardised support for MIHO009 (2g switch)
This is now listed as a registered device class, so it can be used
as part of setup_tool and registry.kvs and other means.

whaleygeek committed:
The standard MIHO009 class is now used in test_2g.py
This is so that now the class is developed it is used throughout
both the tester and setup_tool.

whaleygeek committed:
Rebuilt radio_rpi.so based on latest code.
This should resolve any latent 'doesn't work on Pi version x' issues

whaleygeek committed:
Changed incorrect AttributeError to NameError for raw_input detect

whaleygeek committed:
More testing on 2gang light switch

whaleygeek committed:
Made reference to the NodeRed package, in README

Commit:
Fix undefined error in get_apparent_power for MIHO006

whaleygeek committed:
Merge pull request #128 from Dusty-Meg/master

Fix undefined error in get_apparent_power for MIHO006

whaleygeek committed:
Tested 2gang light switch with an FSK interferer.
Basically has no effect on reliability,
but an OOK interferer in the holdoff period does affect it.

whaleygeek committed:
Implemented framework for tx_silence.
Any send function now returns a number indicating how many seconds
are left before a mandatory tx_silence expires. 0 means transmitted.
This is specifically to enforce the harsh timing requirements of the
2gang light switch.

Devices[*].send_message() returns result of air_interface.send()
All Devices[*].action routines return result of send_message
all air_interface send() methods return 0 for now

This currently has no effect on functionality, it is a baby-steps
architectural change.

whaleygeek committed:
Added a MockSwitch to aid testing.
Added tx_silence wait loops into test_2g.py.
Refactored the switching in test_2g.py to remove repeated code.
Tested MockSwitch in setup_tool and with test_2g

whaleygeek committed:
First algorithm attempt for tx_silence.
Observed that a MIHO009 now forces tx_silence of 3s after any switching.
However, a ENER002 that transmits then allows a MIHO009 immediate switch.
We might need to change this from a 'silence after' to a 'silence before'
to get all the corner cases to pass as well.

whaleygeek committed:
Moved the tx_silence to a pre- barrier rather than a post-.
This works much better as a whole system.
The switching performance is improved throughout, and the reliability
seems rock solid.
Also added a 0.1s tx_pre_silence to all OOKSwitch devices,
as this removes the need to have delays in the application.

whaleygeek committed:
Tested 2gang switch for 1 hour with 3 seconds pre silence
and the given burst parameters. No adverse switching effects observed.

whaleygeek committed:
Saw very occasional switching glitch, moved to 3 second
This is in line with the user documentation as well.

whaleygeek committed:
More bench testing of timing

whaleygeek committed:
More bench testing with 2gang, 1gang and dimmer.
Best parameters to hit all seems to be:
600ms tx, 800ms recharge, 5 repeats, 3 seconds silence at end.
This gives a switching time of about 10 seconds in all cases,
and allows the same parameters to be used with all 3 devices.

whaleygeek committed:
Fixed a number of bugs to do with devices that don't have switches

whaleygeek committed:
Added link to new pico-energenie repo
tidied .gitignore

whaleygeek committed:
Fixed link
